packages/components/input/src/input.css

```diff
@@ -2,36 +2,115 @@
 
 :host,
 :host > div {
+	position: relative;
 	width: 100%;
+	margin-bottom: var(--ts-unit-half);
 }
 
-::slotted(input) {
-	height: var(--ts-unit-plus);
+::slotted(input),
+::slotted(textarea) {
 	width: 100%;
 	font-size: var(--ts-fontsize-mini);
 	font-weight: var(--ts-fontweight);
-	padding: calc(var(--ts-unit-quarter) - 1px) var(--ts-unit-half);
+	padding: var(--ts-unit-quarter) var(--ts-unit-half);
 	border: var(--ts-border-width) var(--ts-border-type) var(--ts-border-color-default);
 	border-radius: var(--ts-radius);
 	background-color: var(--ts-color-white);
+	box-sizing: border-box;
+}
+
+::slotted(input) {
+	height: var(--ts-unit-plus);
 	white-space: nowrap;
 	word-wrap: normal;
 	overflow: hidden;
 	text-overflow: ellipsis;
-	box-sizing: border-box;
 }
 
-::slotted(input:not([disabled]):hover) {
-	border-color: var(--ts-color-blue-light);
-	box-shadow: 0 0 0 1px var(--ts-color-blue-lighter);
+::slotted(textarea) {
+	resize: none;
+	min-height: calc(3 * var(--ts-unit) + var(--ts-unit-half));
+	max-height: calc(10 * var(--ts-unit) + var(--ts-unit-half));
 }
 
-::slotted(input:not([disabled]):focus) {
-	border-color: var(--ts-border-color-focus);
+::slotted(input:not([disabled]):not([readonly]):hover),
+::slotted(textarea:not([disabled]):not([readonly]):hover) {
+	border-color: var(--ts-color-blue-lighter);
+	box-shadow: 0 0 0 1px var(--ts-color-blue-light);
+}
+
+::slotted(input:not([disabled]):not([readonly]):focus-visible),
+::slotted(textarea:not([disabled]):not([readonly]):focus-visible) {
+	outline: none;
+}
+
+::slotted(input:not([disabled]):not([readonly]):focus),
+::slotted(textarea:not([disabled]):not([readonly]):focus) {
+	border-color: var(--ts-color-blue);
+	box-shadow: 0 0 0 1px var(--ts-color-blue-light);
+}
+
+:host([has-error]) {
+	& ::slotted(input:not([disabled]):not([readonly])),
+	& ::slotted(textarea:not([disabled]):not([readonly])) {
+		border-color: var(--ts-color-red);
+	}
+
+	& ::slotted(input:not([disabled]):not([readonly]):hover),
+	& ::slotted(textarea:not([disabled]):not([readonly]):hover) {
+		box-shadow: 0 0 0 1px var(--ts-color-red-light);
+	}
+
+	& ::slotted(input:not([disabled]):not([readonly]):focus),
+	& ::slotted(textarea:not([disabled]):not([readonly]):focus) {
+		box-shadow: 0 0 0 1px var(--ts-color-red-dark);
+	}
 }
 
 /* Disable........................................................ */
 
 ::slotted([disabled]) {
 	color: var(--ts-disabled-color);
 }
+
+/* With icons ........................................................ */
+
+.icon-right,
+.icon-left {
```

Comment:
I suggest to change naming to "icon-start" and "icon-end" so it would be more clear in case of RTL.

Reply:
I agree, I was between right/left and pre/post just went for it for sake of simpler to understand. I'll change it to the start/end 👍